core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2744,6 +2744,25 @@ private[spark] object Utils extends Logging {
     }
   }
 
+  /**
+   * Check the validity of the given Kubernetes master URL and return the resolved URL.
+   */
+  def checkAndGetK8sMasterUrl(rawMasterURL: String): String = {
+    require(rawMasterURL.startsWith("k8s://"),
+      "Kubernetes master URL must start with k8s://.")
+    val masterWithoutK8sPrefix = rawMasterURL.substring("k8s://".length)
```

Comment:
Can we change this String representation to `URI` to make the below check more robust.

Reply:
Done.

Reply:
With the changes, URLs like `k8s://host:port` is no longer valid as the part `host:port` is parsed as a URI with a `host` scheme. Instead, `k8s:///host:port` should be used. I'm not super familiar with URIs, so I'm not sure if this is desirable. /cc @mccheah @foxish.

Reply:
All of the following should be supported and resolved as follows:

- `k8s://host:port` -> `https://host:port`
- `k8s://https://host:port` -> `https://host:port`
- `k8s://http://host:port` -> `http://host:port`

Think we just need to use whatever code that is necessary to get to this state.

Reply:
Are those URIs set in stone? This seems more readable to me:

```
scala> new URI("k8s:http://foo:1234")
res0: java.net.URI = k8s:http://foo:1234

scala> res0.getScheme
res1: String = k8s

scala> res0.getSchemeSpecificPart
res2: String = http://foo:1234
```

It also allows parsing using the `URI` APIs instead of doing string manipulation.

Reply:
This decision is to match how Mesos works with Zookeeper, as these Strings are valid IIRC: `mesos://zk://host1:2181,host2:2181,host3:2181/mesos`. That's just the precedent we are using, but we can use a different convention.

Reply:
(see https://spark.apache.org/docs/latest/running-on-mesos.html#using-a-mesos-master-url)

Reply:
OK, did some changes to make all supported formats work. See https://github.com/apache/spark/pull/19717/commits/51844cc60753456cdf8c2d34a660c239a3c8aa13.